src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -2007,6 +1942,48 @@ void CheckClosed()
         }
 
         #endregion
+
+        #region SafeColumnRead
+
+        TValue SafeColumnRead<TValue>(Func<TValue> func)
+            => SafeColumnRead(() => new ValueTask<TValue>(func()), false).GetAwaiter().GetResult();
+
+        ValueTask<TValue> SafeColumnRead<TValue>(Func<ValueTask<TValue>> func)
+            => SafeColumnRead(func, true);
+
+        async ValueTask<TValue> SafeColumnRead<TValue>(Func<ValueTask<TValue>> func, bool async)
```

Comment:
While it's a good try to avoid code duplication, it brings more state machines and allocations to perf critical path.

Reply:
I agree. Code duplicated to avoid performance impact in nominal cases.